sql/catalyst/src/main/scala-2.13/org/apache/spark/sql/catalyst/util/GenericArrayDataHelper.scala

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.util
+
+import scala.collection.immutable
+import scala.collection.mutable
+
+object GenericArrayDataHelper {
+
+  // SPARK-39766: Special treatment of `immutable.ArraySeq[Any]` and `mutable.ArraySeq[Any]`
+  // to ensure the scenes similar to `arrayOfAnyAsSeq` in `GenericArrayDataBenchmark` have
+  // the same performance when using Scala 2.12 and Scala 2.13
+  def toArray(seq: scala.collection.Seq[Any]): Array[Any] = seq match {
+    case ias: immutable.ArraySeq.ofRef[_] => ias.unsafeArray.asInstanceOf[Array[Any]]
```

Comment:
This is the risky thing, if we are not copying the array behind it. Is this necessary for perf?

Reply:
Do you mean `immutable.ArraySeq` or `mutable.ArraySeq` in particular?

Reply:
If I think it is risky, I can remove this line, because the `WrappedArray` in Scala 2.12 corresponds to `mutable.ArraySeq` in scala 2.13.

And the following conclusions are obtained in https://github.com/scala/bug/issues/12618: https://github.com/scala/bug/issues/12618#issuecomment-1184467529

And I think we can also revert `mutable.ArraySeq` relevant chage when Scala 2.13 fixes it.

Reply:
The latest pr has removed the match for `immutable.ArraySeq`

Reply:
> This is the risky thing, if we are not copying the array behind it. Is this necessary for perf?

When the input type is `mutable.WrappedArray$ofRef` in Scala 2.12,  the result of `toArray` also has no copy:

``` scala
scala> val a = Array[Any](0,"1"); a eq ((a: collection.Seq[Any]).toArray)
a: Array[Any] = Array(0, 1)
res2: Boolean = true
```

Reply:
I keep the changes of `immutable.ArraySeq.ofRef[_]`

Reply:
I'm wondering if we know that the backing array contains the exact contents of the ArraySeq - does it ever have more elements to allow growth? that kind of thing. Maybe WrappedArray does guarantee that. This part seems risky. We could copy; does that hurt perf?

Reply:
> We could copy; does that hurt perf?

Yes, it will affect the performance.  this PR, the throughput of `arrayOfAnyAsSeq` scenario is `41.4M/s` before this pr, and its throughput is `2491.6M/s` after this pr and this performance data is same with using Scala 2.12

Reply:
> I'm wondering if we know that the backing array contains the exact contents of the ArraySeq - does it ever have more elements to allow growth? that kind of thing. Maybe WrappedArray does guarantee that. This part seems risky.

Growth will not be allowed, and the addition or deletion item of `ArraySeq(both mutable and immutable)` will return a new `ArraySeq`,  but there is also a risk: `mutable.ArraySeq.sortInPlace` will sort on the original array,  which is a new method in Scala 2.13 and Spark doesn't use it.

Reply:
Heh yeah I'm still a little worred about using `unsafeArray` but the docs don't suggest it may contain elements not actually in the ArraySeq: "The wrapped mutable Array that backs this ArraySeq. Any changes to this array will break the expected immutability. Its element type does not have to be equal to the element type of this ArraySeq. A primitive ArraySeq can be backed by an array of boxed values and a reference ArraySeq can be backed by an array of a supertype or subtype of the element type."

Reply:
OK, let's wait for upgrade new Scala 2.13.x to solve this issue. I'll close PR first